torch/_custom_op.py

```diff
@@ -0,0 +1,292 @@
+import weakref
+from typing import Callable, Tuple
+
+from torchgen.model import FunctionSchema, SchemaKind
+
+import torch
+import torch._C as _C
+import torch.library as library
+import torch.utils._pytree as pytree
+
+"""
+There are various APIs for defining custom-operator-like things in PyTorch:
+- [user-facing] autograd.Function (Python)
+- [user-facing] CustomOp (Python)
+- [for power users] torch.library (Python)
+- [for power users] TORCH_LIBRARY (C++)
+
+This file contains the implementation for a Simple Custom Operator API (CustomOp).
+Using CustomOp, you are able to define a custom operator and implement interactions
+between the CustomOp and various PyTorch subsystems, including all the subsystems
+that are necessary for a custom operator to work with torch.compile (i.e.,
+autograd, meta, functionalization).
+
+CustomOp is positioned as being safer and easier to use than
+torch.library/TORCH_LIBRARY, which require deep understanding of PyTorch internals,
+as well as being more comprehensive than autograd.Function, which only supports
+implementing gradient computation and vmap rules.
+"""
+
+
+SUPPORTED_DEVICE_TYPES = ("cpu", "cuda")
+
+DEVICE_TYPE_TO_KEY = {
+    "cpu": "CPU",
+    "cuda": "CUDA",
+}
+
+
+# We will not let users register CustomOps with anything that could look like
+# PyTorch internals to avoid confusion.
+RESERVED_NS = {
+    "prims",
```

Comment:
You need `prim` too. But I also think we should not let users pick namespaces and instead base them off of Python module hierarchy.

Reply:
Inferring based off of python module hierarchy sounds pretty good (and fits the theme of "CustomOp" is a layer of abstraction above `torch.Library` with fewer toggles/knobs)